.github/workflows/build_and_publish.yml

```diff
@@ -0,0 +1,115 @@
+# https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows
+
+name: Publish Python 🐍 distribution 📦 to PyPI and TestPyPI
+
+on:
+  push:
+    branches: ['main']
+
+jobs:
+  build:
+    name: Build distribution 📦
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v4
+    - name: Set up Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: "3.x"
+    - name: Install pypa/build
+      run: python3 -m pip install build --user
+    - name: Build a binary wheel and a source tarball
+      run: python3 -m build nipoppy_cli
+    - name: Store the distribution packages
+      uses: actions/upload-artifact@v4
+      with:
+        name: python-package-distributions
+        path: nipoppy_cli/dist/
+
+  publish-to-testpypi:
+    name: Publish Python 🐍 distribution 📦 to TestPyPI
+    needs:
+    - build
+    runs-on: ubuntu-latest
+
+    environment:
+      name: testpypi
+      url: https://test.pypi.org/p/nipoppy
+
+    permissions:
+      id-token: write  # IMPORTANT: mandatory for trusted publishing
+
+    steps:
+    - name: Download all the dists
+      uses: actions/download-artifact@v4
+      with:
+        name: python-package-distributions
+        path: dist/
+    - name: Publish distribution 📦 to TestPyPI
+      uses: pypa/gh-action-pypi-publish@release/v1
+      with:
+        repository-url: https://test.pypi.org/legacy/
```

Comment:
does this work on your main branch @michellewang ?

Reply:
@Remi-Gau I just tried (thanks for the suggestion, didn't think of that), and yes I think so: https://test.pypi.org/project/nipoppy/0.2.0/

Though the README images are not showing up so I will try to fix that if possible. Also will add a badge for PyPI version

Reply:
Ok so because the version is hardcoded right now in `pyproject.toml` I get an error when doing a second push (since a release already exists). I'm trying to figure out how to do the dynamic versioning thing but might need some guidance...

Reply:
will implement something with hatchling

- could help with #309

Reply:
fixed in https://github.com/nipoppy/nipoppy/pull/308/commits/695085d2ce210cd32ee8e1eda9b0835102002c2b

upon pip install of the package you get a `nipoppy_cli/nipoppy/_version.py` module with this type of content

the version number is based on on the git tag


```python
# file generated by setuptools_scm
# don't change, don't track in version control
TYPE_CHECKING = False
if TYPE_CHECKING:
    from typing import Tuple, Union
    VERSION_TUPLE = Tuple[Union[int, str], ...]
else:
    VERSION_TUPLE = object

version: str
__version__: str
__version_tuple__: VERSION_TUPLE
version_tuple: VERSION_TUPLE

__version__ = version = '0.1.1.dev84+gce9c27b.d20240802'
__version_tuple__ = version_tuple = (0, 1, 1, 'dev84', 'gce9c27b.d20240802')
```

So after this

```
git tag 0.2.0
git checkout 0.2.0
pip install .
```

It looks like this

```python
# file generated by setuptools_scm
# don't change, don't track in version control
TYPE_CHECKING = False
if TYPE_CHECKING:
    from typing import Tuple, Union
    VERSION_TUPLE = Tuple[Union[int, str], ...]
else:
    VERSION_TUPLE = object

version: str
__version__: str
__version_tuple__: VERSION_TUPLE
version_tuple: VERSION_TUPLE

__version__ = version = '0.2.0'
__version_tuple__ = version_tuple = (0, 2, 0)
```

Reply:
So on my main I tried:

```bash
git tag 0.3.0
git commit --allow-empty -m 'dummy commit'
git push --tags
git push
```

And I got this
https://github.com/Remi-Gau/nipoppy/actions/runs/10211285451/job/28252341418

It fails now because my fork has no token though.